### PR TITLE
[ROAD-957] Added option for custom CLI paths and auto-update disables

### PR DESCRIPTION
**Pull request by asaf92**
### Description

Added options in settings to disable CLI auto-update and choose a custom CLI path

### Checklist
- [x] CHANGELOG.md updated

### Screenshots / GIFs

image

### CODE REVIEW

File: Snyk.VisualStudio.Extension.Shared/CLI/SnykCli.cs [hunk 1]
```diff
@@ -53,13 +56,21 @@ public ISnykOptions Options
...
+        }
 
         /// <summary>
-        /// Check is CLI file exists in $UserDirectory\.AppData\Snyk.
+        /// Check is CLI file exists.
         /// </summary>
+        /// <param name="customCliPath">Custom CLI path. If null or empty, the default path from AppData will be used.</param>
         /// <returns>True if CLI file exists.</returns>
-        public static bool IsCliExists() => File.Exists(GetSnykCliPath());
+        public static bool DoesCliExist(string customCliPath)
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/CLI/SnykCli.cs:69]: Refactor it to not be public (single source of truth)

File: Snyk.VisualStudio.Extension.Shared/Settings/ISnykOptions.cs [hunk 2]
```diff
@@ -77,6 +87,12 @@ public interface ISnykOptions
         /// <returns><see cref="Task"/> representing the asynchronous operation.</returns>
         Task<bool> IsScanAllProjectsAsync();
 
+        /// <summary>
+        /// Call CLI auth for user authentication at Snyk and get user api token.
+        /// </summary>
+        /// <returns>Returns true if authenticated successfully, false otherwise.</returns>
+        bool Authenticate();
+
         /// <summary>
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/Settings/ISnykOptions.cs:96]: Remove the old `Authenticate`

File: Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/MessagePanel.xaml.cs [hunk 3]
```diff
@@ -95,50 +98,32 @@ private void ShowPanel(StackPanel panel)
...
-            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                authenticationSucceeded = this.ServiceProvider.Options.Authenticate();
+            }
+            catch (FileNotFoundException)
             {
                 await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                this.authenticateSnykProgressBar.Visibility = Visibility.Collapsed;
+                this.Text = "Snyk CLI not found. You can specify a path to a Snyk CLI executable from the settings.";
+                return;
+            }
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/MessagePanel.xaml.cs:119]: Consider catching `Exception`

File: Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/MessagePanel.xaml.cs [hunk 4]
```diff
@@ -95,50 +98,32 @@ private void ShowPanel(StackPanel panel)
             panel.Visibility = Visibility.Visible;
         }
 
-        private void TestCodeNow_Click(object sender, RoutedEventArgs e)
+        private async void TestCodeNow_Click(object sender, RoutedEventArgs e)
```
Comment (asaf92) [inline Snyk.VisualStudio.Extension.Shared/UI/Toolwindow/MessagePanel.xaml.cs:101]: What happens if:
1. Auto-update = false
2. CLI custom path is empty
3. CLI executable is missing

Will we see this dialog?

File: Snyk.VisualStudio.Extension.2022/Snyk.VisualStudio.Extension.2022.csproj [hunk 5]
```diff
@@ -129,4 +130,26 @@
   <Target Name="AfterBuild">
   </Target>
   -->
+  <Target Name="SaveSettingsJsonFile"
```
Comment (michelkaporin) [inline Snyk.VisualStudio.Extension.2022/Snyk.VisualStudio.Extension.2022.csproj:133]: Do you need the same for non-2022?

Comment (asaf92) [reply Snyk.VisualStudio.Extension.2022/Snyk.VisualStudio.Extension.2022.csproj:133]: Absolutely yes.

File: Snyk.VisualStudio.Extension.Shared/Settings/ISnykOptions.cs [hunk 6]
```diff
@@ -63,6 +63,16 @@ public interface ISnykOptions
         /// </summary>
         bool UsageAnalyticsEnabled { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether the CLI should be automatically updated.
+        /// </summary>
+        bool CliAutoUpdate { get; set; }
```
Comment (michelkaporin) [inline Snyk.VisualStudio.Extension.Shared/Settings/ISnykOptions.cs:69]: shouldn't it be `DependenciesAutoUpdate` after the last UI changes?

File: Snyk.VisualStudio.Extension.Shared/Settings/SnykUserStorageSettingsService.cs [hunk 7]
```diff
@@ -29,6 +29,28 @@ public SnykUserStorageSettingsService(string settingsPath, ISnykServiceProvider
             this.settingsLoader = new SnykSettingsLoader(settingsPath);
         }
 
+        public bool CliAutoUpdate
```
Comment (michelkaporin) [inline Snyk.VisualStudio.Extension.Shared/Settings/SnykUserStorageSettingsService.cs:32]: DependenciesAutoUpdate?

File: Snyk.VisualStudio.Extension.Shared/CLI/Download/SnykCliDownloader.cs [hunk 8]
```diff
@@ -400,6 +402,6 @@ private int CliVersionAsInt(string cliVersion)
             }
         }
 
-        private string GetCliFilePath(string filePath) => string.IsNullOrEmpty(filePath) ? SnykCli.GetSnykCliPath() : filePath;
+        private string GetCliFilePath(string filePath) => string.IsNullOrEmpty(filePath) ? SnykCli.GetSnykCliDefaultPath() : filePath;
```
Comment (michelkaporin) [inline Snyk.VisualStudio.Extension.Shared/CLI/Download/SnykCliDownloader.cs:405]: Shouldn't this respect the custom path set in the directory?

**Review by michelkaporin — COMMENTED** [review]
I know it's hard to write tests in the current codebase with the logic scattered across different places, but I encourage you to try to refactor it out going forward to make it unit testable and test main logic pieces.

Without having tests it's difficult to prove that it works as expected. Something to test would be:
- Does download happen when "dependenciesAutoUpdate" is false?
- Does download happen to the custom path provided in the settings?
- What happens if download path isn't provided in settings?
- What happens if invalid path is provided?

File: Snyk.VisualStudio.Extension.Shared/CLI/Download/SnykCliDownloader.cs (hunk 8 above)
Comment (asaf92) [reply Snyk.VisualStudio.Extension.Shared/CLI/Download/SnykCliDownloader.cs:405]: It does, that's the argument. I'll add docs to clarify it.

Comment (michelkaporin) [reply Snyk.VisualStudio.Extension.Shared/CLI/Download/SnykCliDownloader.cs]: Ah right. What confused me is that the logic of obtaining the custom path lies in TasksService and not here. It's another point to refactoring. Renaming function argument from `filePath` to `customCliPath` helped to comprehend it.

**Pull request merged**